oozie-to-airflow/tests/test_templates.py

```diff
@@ -0,0 +1,391 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests Templates"""
+import ast
+import unittest
+from copy import deepcopy
+from random import randint
+from typing import Dict, Any
+
+from parameterized import parameterized
+from airflow.utils.trigger_rule import TriggerRule
+
+from converter.primitives import Task, Relation
+from utils.template_utils import render_template
+
+DELETE_MARKER: Any = {}
+
+
+def mutate(parent: Dict[str, Any], mutations: Dict[str, Any]):
+    result = {}
+    result.update(parent)
+    for key, value in mutations.items():
+        if value is DELETE_MARKER:
+            del result[key]
+        elif isinstance(value, dict):
+            result[key] = mutate(parent.get(key, {}), value)
+        else:
+            result[key] = value
+    return result
+
+
+def get_by_path(target, segments):
+    result = target
+    for segment in segments:
+        if isinstance(result, dict):
+            result = result[segment]
+        elif isinstance(result, list):
+            result = result[int(segment)]
+        else:
+            raise Exception(f"Invalid path: {'.'.join(segments)}")
+    return result
+
+
+def set_by_path(target, segments, value):
+    result = get_by_path(target, segments[:-1])
+    if isinstance(result, dict):
+        result[segments[-1]] = value
+    elif isinstance(result, list):
+        result[int(segments[-1])] = value
+    else:
+        raise Exception(f"Invalid path: {'.'.join(segments)}")
+
+
+class TemplateTestMixin:
+    # pylint: disable=invalid-name
+    def assertValidPython(self, code):
```

Comment:
Is there any particular reason for this not being `assert_valid_python()`?

Reply:
The Python convention assumes that the methods that perform the assertions have such a style. 
https://docs.python.org/3/library/unittest.html
I preferred to avoid several styles of naming functions intended for one role.  This forces thinking of what convention has been used in this particular case.

Reply:
Ok, cool.